oracle/oracles/pd.go

```diff
@@ -366,3 +638,84 @@ func (o *pdOracle) SetExternalTimestamp(ctx context.Context, ts uint64) error {
 func (o *pdOracle) GetExternalTimestamp(ctx context.Context) (uint64, error) {
 	return o.c.GetExternalTimestamp(ctx)
 }
+
+func (o *pdOracle) getCurrentTSForValidation(ctx context.Context, opt *oracle.Option) (uint64, error) {
+	ch := o.tsForValidation.DoChan(opt.TxnScope, func() (interface{}, error) {
+		metrics.TiKVValidateReadTSFromPDCount.Inc()
+
+		// If the call that triggers the execution of this function is canceled by the context, other calls that are
+		// waiting for reusing the same result should not be canceled. So pass context.Background() instead of the
+		// current ctx.
+		res, err := o.GetTimestamp(context.Background(), opt)
+		return res, err
+	})
+	select {
+	case <-ctx.Done():
+		return 0, errors.WithStack(ctx.Err())
+	case res := <-ch:
+		if res.Err != nil {
+			return 0, errors.WithStack(res.Err)
+		}
+		return res.Val.(uint64), nil
+	}
+}
+
+func (o *pdOracle) ValidateSnapshotReadTS(ctx context.Context, readTS uint64, opt *oracle.Option) error {
+	latestTS, err := o.GetLowResolutionTimestamp(ctx, opt)
+	// If we fail to get latestTS or the readTS exceeds it, get a timestamp from PD to double-check.
+	// But we don't need to strictly fetch the latest TS. So if there are already concurrent calls to this function
+	// loading the latest TS, we can just reuse the same result to avoid too many concurrent GetTS calls.
```

Comment:
The current implementation is enough I think. This is just a minor suggestion. 

Use singleflight to reduce the GetTS calls is reasonable, but it's possible that a valid ts is invalid in this case.

1. thread 1 validate current_ts and send a get ts request, PD allocates its latest TSO and the response is slow.
2. thread 2 validate current_ts and the singleflight will reuse the thread 1's TSO.
3. The GetTS request of thread 1's is responded.

Then thread 2 can get a stale TSO from PD which increases the failure possibility.

Maybe an enhancement is to resend a singleflight GetTS request if `readTS > latestTS` to recheck the timestamp in this case.